src/samplers/cpu/mod.rs

```diff
@@ -150,6 +128,60 @@ impl Sampler for Cpu {
 }
 
 impl Cpu {
+    #[cfg(feature = "bpf")]
+    fn initialize_bpf_perf(&mut self) -> Result<(), std::io::Error> {
+        let cpus = crate::common::hardware_threads().unwrap();
+        let interval = self
+            .sampler_config()
+            .interval()
+            .unwrap_or_else(|| self.general_config().interval()) as u64;
+        let frequency = if interval > 1000 {
+            1
+        } else if interval == 0 {
+            1
+        } else {
+            1000 / interval
+        };
+
+        let code = format!(
+            "{}\n{}",
+            format!("#define NUM_CPU {}", cpus),
+            include_str!("perf.c").to_string()
+        );
+        if let Ok(mut bpf) = bcc::BPF::new(&code) {
+            for statistic in self.sampler_config().statistics().iter() {
+                if let Some(table) = statistic.table() {
+                    if let Some(event) = statistic.event() {
+                        if PerfEventArray::new()
+                            .table(&format!("{}_array", table))
+                            .event(event)
+                            .attach(&mut bpf)
+                            .is_err()
+                        {
+                            if !self.common().config().general().fault_tolerant() {
+                                fatal!("failed to initialize perf bpf for event: {:?}", event);
+                            } else {
+                                error!("failed to initialize perf bpf for event: {:?}", event);
+                            }
+                        }
+                    }
+                }
+            }
+            PerfEvent::new()
+                .handler("do_count")
+                .event(Event::Software(SoftwareEvent::CpuClock))
+                .sample_frequency(Some(frequency))
+                .attach(&mut bpf)
```

Comment:
I think it is possible for bpf to fail here as well?

Reply:
Fixed